### PR TITLE
Fix multi-alarm overlapping bug

### PR DESCRIPTION
**Pull request by alevy**
This PR fixes a subtle bug in the userspace virtual alarm library. If alarms are created and expire pretty close to each other, the `within_range` check can incorrectly return the relative relationship between the two, place the earlier alarm _after_ the later alarm, which is incorrect, and can end up totally starving the earlier alarm if the later alarm is repeating.

This bug was introduced when trying to correctly account for potentially overflowing alarms, and the "obvious" fix (just compare absolute expiration values) doesn't work.

This PR adds two tests, one that demonstrates the starved alarm issue and one that exercises the overflow issue, and fixes the bug in such a way that both pass. In particular, the fix is simply to explicitly check if alarms overflow when insert them to the virtual alarm queue and use that for the comparison between alarms.

## Test 1: Multi-Alarm Simple

This is a simple test for multiple alarms that only sets exactly two repeating alarms that should overlap frequently (one is twice the other's frequency). It sets two repeating alarms, at 500ms and 1s respectively, each prints the alarm index (1 or 2), current tick and alarm expiration to the console. When successful, both alarms should fire and alarm 1 should fire twice as often as alarm 2.

Today (on the NRF52840 at least, though I don't expect other chips to be different), this test fails. It should output (the first column in each line is the timer index):

```sh
1 6316032 6314240
2 10512384 10510592
1 10512384 10511104
1 14722560 14720768
2 18903552 18901760
1 18919680 18917632
1 23116544 23114752
2 27294720 27292928
...
```

but, due to a bug in the alarm library, actually outputs:

```sh
1 7254784 7252992
2 11451136 11449344
2 19842304 19840512
2 28233472 28231680
2 36624640 36622848
2 45015808 45014016
2 53406976 53405184
...
```

(note timer `1` never appears after the first instance)

## Test 2: Multi-Alarm Overflow

This is a for multiple alarms where one overflows the clock (`UINT32_MAX` ticks), resulting in a low absolute value expiration, while the other is "normal" (1 second). When successful, both alarms should fire, the second after about 1 second, and the first after the clock overflows (around 7 minutes on 32kHz clocks).

When the logic for inserting clocks doesn't account correctly for overflows, the second alarm fires _after_ the first one, both after the clock overflows.

### CODE REVIEW

**Review by tyler-potyondy — DISMISSED** [review]
This is an interesting bug and perhaps explains some of the edge case bugs we've seen with the OpenThread library. 

I'm curious to see what the root cause ends up being.

File: examples/tests/multi_alarm_simple_test/main.c [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <libtock-sync/services/alarm.h>
+
+static void event_cb(__attribute__ ((unused)) uint32_t now,
+                     __attribute__ ((unused)) uint32_t expiration,
+                     void*                             ud) {
```
Comment (ppannuto) [inline examples/tests/multi_alarm_simple_test/main.c:8]:
```suggestion
static void event_cb(uint32_t now, uint32_t expiration, void* ud) {
```

I'm mildly annoyed that using an "unused"isn't some kind of warning...

**Review by brghena — APPROVED** [review]

**Comment by lschuermann** [comment]
Good catch, and the test seems fine to me.

However, this does raise a more general issue to me, and one that relates to our releases, automated tests, and the timer subsystem in particular: how to assert whether the test ran correctly, or not.

In particular, for timer tests we often just produce some raw, unprocessed output. When it comes to running these tests as part of a release (esp. with our release cadence), we then need to re-learn what it means for these tests to pass or fail. Similarly, with automated tests, esp. with non-deterministic output, it can be hard to encode a check function that determines the test result.

We may want to take this opportunity to figure out what we can add to the user space tests and/or a test framework to make this easier.

**Comment by lschuermann** [comment]
FWIW, @charles37 is working on an automated test for this, to be integrated into Treadmill.

**Comment by alevy** [comment]
A significant update that adds another test (for clock overflow) and (I believe) fixes the underlying issue. See new PR description.

**Comment by brghena** [comment]
I will look over this tomorrow. Thanks for the nudge.

**Review by tyler-potyondy — CHANGES_REQUESTED** [review]
The logic for the timer placement seems to check out and is an improvement over the somewhat opaque within range function. I have one small comment for improving the readability of the logic, but overall the changes look good to me.

File: libtock/services/alarm.c [hunk 2]
```diff
@@ -75,19 +70,44 @@ static uint32_t ticks_to_ms(uint32_t ticks) {
...
   while (*cur != NULL) {
+    // Compute the tick value at which this alarm will expire.
     uint32_t cur_expiration = (*cur)->reference + (*cur)->dt;
-    uint32_t new_expiration = alarm->reference + alarm->dt;
-    if (!within_range(alarm->reference, cur_expiration, new_expiration)) {
+    // Determine whether the clock overflows before this alarm
+    // expires.
+    bool cur_overflows = (*cur)->reference > UINT32_MAX - (*cur)->dt;
+
+    // This alarm happens after the new alarm if:
```
Comment (brghena) [inline libtock/services/alarm.c:103]: What is "this alarm" and what is "the new alarm"? I think it's `cur` and `alarm` respectively? But I might have that backwards. In my opinion, this entire comment block should use the actual names for clarity.

File: libtock/services/alarm.c [hunk 3]
```diff
@@ -75,19 +70,44 @@ static uint32_t ticks_to_ms(uint32_t ticks) {
...
+    // - neither expirations overflow and this expiration value is larger than the new expiration
+    // - both overflow and this expiration value is larger than the new expiration
+    // - or, this alarm overflows but the new one doesn't
+    //
+    // If the new alarm overflows and this alarm doesn't, this alarm
+    // happens _before_ the new alarm.
+    if (!(!cur_overflows && new_overflows) && ((cur_overflows && !new_overflows) || cur_expiration > new_expiration)) {
```
[the remark this answers is not in the document]
Comment (brghena) [reply libtock/services/alarm.c:110]: I strongly agree. I really can't parse line 110.

Based on my reading of the comment, the logic would be `if ((cur_overflows && !new_overflows) || (cur_expiration > new_expiration))`. I don't understand why the first part of the condition is there.

File: examples/tests/multi_alarm_simple_overflow_test/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Test Multiple Alarms (With Overflow)
+
+This tests the virtual alarms available to userspace. It sets two
+alarms, first one that overflows the alarm, such that it's expiration
```
Comment (brghena) [inline examples/tests/multi_alarm_simple_overflow_test/README.md:4]:
```suggestion
alarms, first one that overflows the alarm, such that its expiration
```

File: examples/tests/multi_alarm_simple_overflow_test/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+is small in absolute value (but should shouldn't fire until after the
```
Comment (brghena) [inline examples/tests/multi_alarm_simple_overflow_test/README.md:5]: Should or shouldn't? Pick one please

**Review by brghena — CHANGES_REQUESTED** [review]
This is a good catch and the fix looks good to me. I have a bunch of small things to clean up though.

**Review by ppannuto — CHANGES_REQUESTED** [review]
Review of everything except timer logic so far sorry

File: examples/tests/multi_alarm_simple_test/README.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Test Multiple Alarms (Simple)
+
+This tests the virtual alarms available to userspace. It sets two
+repeating alarms, at 500ms and 1s respectively, each prints the alarm
+index (1 or 2), current tick and alarm expiration to the console. When
+successful, both alarms should fire and alarm 1 should fire twice as
+often as alarm 2.
```
[the remark this answers is not in the document]
Comment (alevy) [reply examples/tests/multi_alarm_simple_test/README.md:7]: Done, see below

File: libtock/services/alarm.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (alevy) [reply libtock/services/alarm.c]: I couldn't quite help myself from being a tiny bit clever still, but I believe the new version of both the comment and boolean logic are clear now.

File: examples/tests/multi_alarm_simple_overflow_test/README.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+clock overflows) and one after 1s. When successful, the second (1s)
+alarm should fire after 1 second, while the second alarm should fire
+after the clock overflows (approximately 7 minutes if the clock is at
+32kHz).
+
+If the virtual alarm library is buggy, this test might fail by
+scheduling the 1 second alarm _after_ the longer wrapping alarm, and
+it won't fire immediately.
```
[the remark this answers is not in the document]
Comment (alevy) [reply examples/tests/multi_alarm_simple_overflow_test/README.md:13]: See below

**Comment by alevy** [comment]
@ppannuto @tyler-potyondy @brghena all comments addressed

**Review by brghena — DISMISSED** [review]

**Review by tyler-potyondy — DISMISSED** [review]

**Review by ppannuto — APPROVED** [review]
